lib/src/merge.rs

```diff
@@ -249,53 +249,16 @@ impl<T> Merge<T> {
         }
     }
 
-    /// Returns a vector mapping of a value's index in the simplified merge to
-    /// its original index in the unsimplified merge.
-    ///
-    /// The merge is simplified by removing identical values in add and remove
-    /// values.
-    fn get_simplified_mapping(&self) -> Vec<usize>
-    where
-        T: PartialEq,
-    {
-        let unsimplified_len = self.values.len();
-        let mut simplified_to_original_indices = (0..unsimplified_len).collect_vec();
-
-        let mut add_index = 0;
-        while add_index < simplified_to_original_indices.len() {
-            let add = &self.values[simplified_to_original_indices[add_index]];
-            let mut remove_indices = simplified_to_original_indices
-                .iter()
-                .enumerate()
-                .skip(1)
-                .step_by(2);
-            if let Some((remove_index, _)) = remove_indices
-                .find(|&(_, original_remove_index)| &self.values[*original_remove_index] == add)
-            {
-                // Align the current "add" value to the `remove_index/2`-th diff, then
-                // delete the diff pair.
-                simplified_to_original_indices.swap(remove_index + 1, add_index);
-                simplified_to_original_indices.drain(remove_index..remove_index + 2);
-            } else {
-                add_index += 2;
-            }
-        }
-
-        simplified_to_original_indices
-    }
-
     /// Simplify the merge by joining diffs like A->B and B->C into A->C.
     /// Also drops trivial diffs like A->A.
     pub fn simplify(mut self) -> Self
     where
         T: PartialEq + Clone,
     {
-        let mapping = self.get_simplified_mapping();
-        // Reorder values based on their new indices in the simplified merge.
-        self.values = mapping
-            .iter()
-            .map(|index| self.values[*index].clone())
-            .collect();
+        if self.values.len() == 1 {
+            return self;
+        }
+        self.values = simplify_internal(self.values.to_vec()).into();
```

Comment:
nit: you can use `.into_vec()` to avoid allocation, but maybe we can make `simplify_internal()` not consume the input vec instead.